### ISSUE TITLE
feature request: bolt interactive mode

### ISSUE THREAD

**Issue by scoopex** [puppetlabs/bolt#429]
Currently it is not possible to use bolt for interactive ssh remote sessions.

Example:
```
# continuously analyze the io behavior of a number of remote hosts
bolt command run "vim /etc/puppetlabs/puppet/puppet.conf; puppet agent --test" -n host1,host2,host3

# perform distribution updates
bolt command run "do-release-upgrade && reboot" -n host1,host2,host3
```

Comparable to https://github.com/scoopex/hostctl (-l, -p) it woule be nice to have three additional feature-options:

* --login: just invoke a login to the remote machine in single threaded mode with direct stdin/stdou/stderr
* --interactive : execute the specified script/command in single threaded mode with direct stdin/stdou/stderr
* --prompt:  can be used in combination with --interactive, invoke the command and ask for (c)ontinue to next host, (r)etry command, (s)hell invocation, (q)uit after the execution

**Comment by LongLiveCHIEF** [comment]
Just my opinion here (and I'm not a collaborator/member of this project), but I think this is out of scope for bolt. There are tools out there to do this, and it seems contradictory to the nature of puppet (which bolt is designed to be utilized with).

Puppet is for automation, and bolt is a way to trigger that automation. If you've gotten to the point of needing interactions and prompts, you're no longer talking automation, and that feels like it no longer belongs.

**Comment by MikaelSmith** [comment]
I agree that this seems fairly different from the use-cases Bolt is targeting. I don't see an argument for why adding this to Bolt would be any better than just using hostctl or similar tools.

**Comment by scoopex** [comment]
My intention for this feature request was that the described helps for the following situations:

- if something went wrong with a number of hosts, this simplifies the problem inspection
- i often perform canary tests to find problems with automation code, after a number of hosts i can change to unattended execution
- this helps to manage critical automation tasks (like distribution upgrades) in a half-automated way

I really love bolt, and these feature seems to be the last remaining part to eliminate all the insufficient cluster-execution / cluster-shell tools i used in the last years

**Comment by scoopex** [comment]
Probably it would be a good idea to split this three parts:

- #102 - streaming of stderr/stdout output
- support for canary runs with user interaction with : (c)ontinue to next host, (r)etry command, (s)hell invocation, (q)uit
- support for non paralleled interactive shell sessions with interaction over stdin, stderr, stdout

What do you think?

**Comment by MikaelSmith** [comment]
(1) seems doable. (2) I think continue, retry, quit make sense, shell invocation starts to feel like a substantial investment in a new direction, which (3) continues. I can see how interactive shells using inventory would be valuable, but I don't see my team going in that direction any time soon. Maybe we could get there by invoking a separate tool to handle the interactive session?
